File: server/scripts/generate_app.dart
import 'dart:io';
import 'package:path/path.dart' as path;
import 'package:yaml/yaml.dart';

class AppGenerator {
  final String templatePath;
  final String outputPath;
  final Map<String, dynamic> config;

  AppGenerator({
    required this.templatePath,
    required this.outputPath,
    required this.config,
  });

  Future<void> generate() async {
    try {
      // Create output directory
      await Directory(outputPath).create(recursive: true);

      // Copy template files
      await _copyTemplateFiles();

      // Update pubspec.yaml
      await _updatePubspec();

      // Update app configuration
      await _updateAppConfig();

      // Generate platform-specific files
      await _generatePlatformFiles();

      print('App generated successfully at: $outputPath');
    } catch (e) {
      print('Error generating app: $e');
      rethrow;
    }
  }

  Future<void> _copyTemplateFiles() async {
    final templateDir = Directory(templatePath);
    await for (final entity in templateDir.list(recursive: true)) {
      if (entity is File) {
        final relativePath = path.relative(entity.path, from: templatePath);
        final targetPath = path.join(outputPath, relativePath);
        await entity.copy(targetPath);
      }
    }
  }

  Future<void> _updatePubspec() async {
    final pubspecPath = path.join(outputPath, 'pubspec.yaml');
    final pubspecFile = File(pubspecPath);
    final pubspecContent = await pubspecFile.readAsString();
    final pubspec = loadYaml(pubspecContent);

    // Update app name and description
    pubspec['name'] = config['app_name'];
    pubspec['description'] = config['description'];

    // Update dependencies based on features
    if (config['features']['notifications']) {
      pubspec['dependencies']['flutter_local_notifications'] = '^16.3.0';
    }
    if (config['features']['gps']) {
      pubspec['dependencies']['google_maps_flutter'] = '^2.5.3';
    }
    // Add other feature-specific dependencies...

    await pubspecFile.writeAsString(pubspec.toString());
  }

  Future<void> _updateAppConfig() async {
    final configPath = path.join(outputPath, 'lib/config/app_config.dart');
    final configFile = File(configPath);

    final configContent = '''
// Generated app configuration
class AppConfig {
  static const String appName = '${config['app_name']}';
  static const String appVersion = '${config['version']}';
  
  // Feature flags
  static const bool hasNotifications = ${config['features']['notifications']};
  static const bool hasCoupons = ${config['features']['coupons']};
  static const bool hasGps = ${config['features']['gps']};
  static const bool hasPhotoReporting = ${config['features']['photo_reporting']};
  static const bool hasSocialSharing = ${config['features']['social_sharing']};
  
  // Theme configuration
  static const Map<String, String> colors = ${config['customization']['colors']};
  static const Map<String, String> fonts = ${config['customization']['fonts']};
}
''';

    await configFile.writeAsString(configContent);
  }

  Future<void> _generatePlatformFiles() async {
    // Generate Android files
    if (config['platforms']['android']) {
      await _generateAndroidFiles();
    }

    // Generate iOS files
    if (config['platforms']['ios']) {
      await _generateIOSFiles();
    }
  }

  Future<void> _generateAndroidFiles() async {
    final androidPath = path.join(outputPath, 'android');

    // Update AndroidManifest.xml
    final manifestPath = path.join(
      androidPath,
      'app/src/main/AndroidManifest.xml',
    );
    final manifestFile = File(manifestPath);
    var manifestContent = await manifestFile.readAsString();

    // Add permissions based on features
    if (config['features']['gps']) {
      manifestContent = manifestContent.replaceFirst(
        '</manifest>',
        '    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />\n    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />\n</manifest>',
      );
    }

    await manifestFile.writeAsString(manifestContent);
  }

  Future<void> _generateIOSFiles() async {
    final iosPath = path.join(outputPath, 'ios');

    // Update Info.plist
    final plistPath = path.join(iosPath, 'Runner/Info.plist');
    final plistFile = File(plistPath);
    var plistContent = await plistFile.readAsString();

    // Add permissions based on features
    if (config['features']['gps']) {
      plistContent = plistContent.replaceFirst(
        '</dict>',
        '    <key>NSLocationWhenInUseUsageDescription</key>\n    <string>This app needs access to location when open to show your position on the map.</string>\n    <key>NSLocationAlwaysUsageDescription</key>\n    <string>This app needs access to location when in the background to show your position on the map.</string>\n</dict>',
      );
    }

    await plistFile.writeAsString(plistContent);
  }
}

File: server/scripts/generate_app.js
const fs = require('fs-extra');
const path = require('path');
const { execSync } = require('child_process');
const { v4: uuidv4 } = require('uuid');

class AppGenerator {
  constructor(templateDir, buildsDir) {
    this.templateDir = templateDir;
    this.buildsDir = buildsDir;
  }

  async generateApp(config) {
    try {
      // Create unique build directory
      const buildId = uuidv4();
      const buildDir = path.join(this.buildsDir, buildId);
      console.log('Creating app in directory:', buildDir);

      // Copy template
      await fs.copy(this.templateDir, buildDir);
      console.log('Template copied successfully');

      // Update app configuration
      await this._updateAppConfig(buildDir, config);
      console.log('Configuration updated');

      // Update Android configuration
      await this._updateAndroidConfig(buildDir, config);
      console.log('Android configuration updated');

      // Update iOS configuration (if needed)
      // await this._updateIOSConfig(buildDir, config);

      // Build the app
      console.log('Building APK...');
      try {
        execSync('flutter build apk --release', { 
          cwd: buildDir,
          stdio: 'inherit'
        });

        const apkPath = path.join(buildDir, 'build', 'app', 'outputs', 'flutter-apk', 'app-release.apk');
        
        // Verify the APK exists
        if (!fs.existsSync(apkPath)) {
          throw new Error('APK file not found after build');
        }

        return {
          success: true,
          buildId,
          apkPath,
          buildDir
        };
      } catch (buildError) {
        console.error('Build error:', buildError);
        return {
          success: false,
          error: 'Build failed: ' + buildError.message,
          buildId,
          buildDir
        };
      }

    } catch (error) {
      console.error('Error generating app:', error);
      return {
        success: false,
        error: error.message,
        details: error.stack
      };
    }
  }

  async _updateAppConfig(buildDir, config) {
    const configPath = path.join(buildDir, 'assets', 'config.json');
    await fs.ensureDir(path.dirname(configPath));
    await fs.writeJson(configPath, config, { spaces: 2 });

    // Update pubspec.yaml with the new app name
    const pubspecPath = path.join(buildDir, 'pubspec.yaml');
    let pubspec = await fs.readFile(pubspecPath, 'utf8');
    pubspec = pubspec.replace(/name: .*/, `name: ${config.appName.toLowerCase().replace(/\s+/g, '_')}`);
    await fs.writeFile(pubspecPath, pubspec);
  }

  async _updateAndroidConfig(buildDir, config) {
    // Update Android package name
    const manifestPath = path.join(buildDir, 'android', 'app', 'src', 'main', 'AndroidManifest.xml');
    let manifest = await fs.readFile(manifestPath, 'utf8');
    manifest = manifest.replace(
      /package="[^"]*"/,
      `package="${config.packageName || 'com.example.app'}"`
    );
    manifest = manifest.replace(
      /android:label="[^"]*"/,
      `android:label="${config.appName}"`
    );
    await fs.writeFile(manifestPath, manifest);

    // Update build.gradle
    const buildGradlePath = path.join(buildDir, 'android', 'app', 'build.gradle');
    let buildGradle = await fs.readFile(buildGradlePath, 'utf8');
    buildGradle = buildGradle.replace(
      /applicationId "[^"]*"/,
      `applicationId "${config.packageName || 'com.example.app'}"`
    );
    await fs.writeFile(buildGradlePath, buildGradle);
  }
}

// Example usage
const generator = new AppGenerator(
  path.join(__dirname, '..', 'templates', 'flutter_basic'),
  path.join(__dirname, '..', 'builds')
);

// If running directly (not imported as a module)
if (require.main === module) {
  const sampleConfig = {
    appName: "Custom App",
    packageName: "com.example.customapp",
    theme: {
      primaryColor: "0xFF2196F3"
    },
    pages: [
      {
        title: "Home",
        type: "list",
        content: {
          items: [
            {
              title: "Custom Item",
              description: "This is a custom item"
            }
          ]
        }
      }
    ],
    settings: {
      version: "1.0.0"
    }
  };

  generator.generateApp(sampleConfig)
    .then(result => {
      if (result.success) {
        console.log('App generated successfully!');
        console.log('APK location:', result.apkPath);
      } else {
        console.error('Failed to generate app:', result.error);
      }
    })
    .catch(error => {
      console.error('Error:', error);
    });
}

module.exports = AppGenerator; 